src/small_list.ml

```diff
@@ -0,0 +1,192 @@
+(*
+ * Copyright (c) 2021 Tarides <contact@tarides.com>
+ *
+ * Permission to use, copy, modify, and distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *)
+
+type 'a t =
+  | Tuple0
+  | Tuple1 of 'a
+  | Tuple2 of 'a * 'a
+  | Tuple3 of 'a * 'a * 'a
+  | Tuple4 of 'a * 'a * 'a * 'a
+  | Tuple5 of 'a * 'a * 'a * 'a * 'a
+  | Tuple6 of 'a * 'a * 'a * 'a * 'a * 'a
```

Comment:
maybe this is obvious, but why up to 6?

Reply:
Not obvious at all; I just got bored after 6 :wink: In practice, sizes above about 7 don't really happen with the hash functions we're using, so it's not worth going any further. If we used `Obj` rather than a type-safe implementation, we could push the threshold much higher & also improve the speed.

As it happens, @pascutto has a WIP patch to use open-addressing in the hashtable, which would avoid the need for `small_list.ml` entirely. For now, this is probably good enough.